File: src/main/java/xyz/cofe/jcommander/FilesTable.java
package xyz.cofe.jcommander;

import org.checkerframework.checker.nullness.qual.EnsuresNonNull;
import org.checkerframework.checker.nullness.qual.MonotonicNonNull;
import org.checkerframework.checker.nullness.qual.NonNull;
import xyz.cofe.collection.EventList;
import xyz.cofe.jcommander.table.Column;
import xyz.cofe.jcommander.table.Table;
import xyz.cofe.jcommander.table.TableColumns;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.attribute.PosixFilePermission;
import java.util.Comparator;
import java.util.Optional;
import java.util.Set;

/**
 * Таблица файлов
 */
public class FilesTable extends Table<Path> {
    private static void log( Throwable err ){
    }

    //region ctor
    public FilesTable( EventList<Path> values ){
        super(values);
    }

    public FilesTable(){
    }
    //endregion

    //region columns

    /**
     * Колонки связанные с файлами, для таблицы
     */
    public static class Columns extends TableColumns<Path> {
        /**
         * Конструктор по умолчанию
         */
        public Columns(){
            this(false);
        }

        /**
         * Конструктор
         * @param includeDefaultColumns true - добавить колонки
         */
        @SuppressWarnings("nullness")
        public Columns(boolean includeDefaultColumns){
            if( includeDefaultColumns ){
                add(baseName);
                add(extension);
                add(unixPermissions);
                add(humanSize);
            }
        }

        private final static Set<String> specialNames = Set.of(".", "..");

        //region baseName : Column
        @SuppressWarnings("nullness")
        public final Column<Path,String> baseName = build(
            p -> specialNames.contains(getName(p))
                ? "/"+getName(p)
                : getBasename(p).length()<1
                    ? getName(p)
                    : Files.isDirectory(p)
                        ? "/" + getBasename(p)
                        : Files.isSymbolicLink(p)
                            ? "~" + getBasename(p)
                            : " " + getBasename(p),
            c -> c.width(20).minMaxWidth(5,100).name("name").fixed(false)
        );
        //endregion
        //region extension : Column
        @SuppressWarnings("nullness")
        public final Column<Path,String> extension = build(
            p -> specialNames.contains(getName(p))
                ? ""
                : getBasename(p).length()<1
                    ? ""
                    : getExtension(p),
            c -> c.name("ext").width(5).minMaxWidth(3,20).fixed(false)
        );
        //endregion
        //region humanSize : Column
        private static Optional<Long> safeSize(Path p){
            try{
                if( Files.exists(p) ){
                    return Optional.of(Files.size(p));
                }else{
                    return Optional.empty();
                }
            } catch( IOException e ){
                log(e);
                return Optional.empty();
            }
        }

        @SuppressWarnings("nullness")
        public final Column<Path,String> humanSize = build(
            p -> Files.exists(p)
                ? safeSize(p).map(FilesTable::humanSize).orElse("?")
                : "-1",
            c -> c.name("size").width(6).minMaxWidth(6,12).fixed(false)
        );
        //endregion
        //region unixPermissions : Column
        private String perm( Set<PosixFilePermission> perms ) {
            //noinspection StringBufferReplaceableByString
            StringBuilder sb = new StringBuilder();
            sb.append( perms.contains(PosixFilePermission.OWNER_READ) ? "r" : "-" );
            sb.append( perms.contains(PosixFilePermission.OWNER_WRITE) ? "w" : "-" );
            sb.append( perms.contains(PosixFilePermission.OWNER_EXECUTE) ? "x" : "-" );
            sb.append( perms.contains(PosixFilePermission.GROUP_READ) ? "r" : "-" );
            sb.append( perms.contains(PosixFilePermission.GROUP_WRITE) ? "w" : "-" );
            sb.append( perms.contains(PosixFilePermission.GROUP_EXECUTE) ? "x" : "-" );
            sb.append( perms.contains(PosixFilePermission.OTHERS_READ) ? "r" : "-" );
            sb.append( perms.contains(PosixFilePermission.OTHERS_WRITE) ? "w" : "-" );
            sb.append( perms.contains(PosixFilePermission.OTHERS_EXECUTE) ? "x" : "-" );
            return sb.toString();
        }
        private Optional<Set<PosixFilePermission>> safeUnixPermissions( Path p ){
            try{
                return Optional.of(Files.getPosixFilePermissions(p));
            } catch( IOException e ){
                log(e);
                return Optional.empty();
            }
        }

        @SuppressWarnings("nullness")
        public final Column<Path,String> unixPermissions = build(
            p -> Files.exists(p)
                ? safeUnixPermissions(p).map(this::perm).orElse("?")
                : "?",
            c -> c.width(9).fixed(true).name("perm")
        );
        //endregion
    }

    private @MonotonicNonNull Columns columns;

    @SuppressWarnings("nullness")
    @Override
    @EnsuresNonNull("this.columns")
    public @NonNull Columns getColumns(){
        if( columns!=null )return columns;
        columns = new Columns(true);
        return columns;
    }
    //endregion

    /**
     * Базовое имя файла без расширения
     * @return базовое имя файла
     */
    private static String getBasename( @NonNull Path path ){
        @SuppressWarnings("nullness")
        @NonNull String name = path.getFileName().toString();
        int dot = name.lastIndexOf(".");
        if( dot<0 )return name;
        if( dot==0 )return "";
        return name.substring(0, dot);
    }

    /**
     * Раширение имени файла.
     * @return расширение имени файла без точки
     */
    private static String getExtension( @NonNull Path path ){
        @SuppressWarnings("nullness")
        @NonNull String name = path.getFileName().toString();
        int dot = name.lastIndexOf(".");
        if( dot<0 )return "";
        if( dot>=(name.length()-1) )return "";
        return name.substring(dot+1);
    }

    /**
     * Имя файла
     * @return Имя файла
     */
    @SuppressWarnings("nullness")
    private static String getName( @NonNull Path path ){
        return path.getFileName().toString();
    }

    /**
     * Человеко читабельный размер файла
     * @param bytes размер в байтах
     * @return читабельный размер
     */
    private static String humanSize( long bytes ){
        String sz = "";
        if( bytes>=(1024*1024*1024) ){
            sz = Long.toString(bytes >> 30)+"g";
        }else if( bytes>=(1024*1024) ){
            sz = Long.toString(bytes >> 20)+"m";
        }else if( bytes>=(1024) ){
            sz = Long.toString(bytes >> 10)+"k";
        }else {
            sz = Long.toString(bytes);
        }

        if( sz.length()<5 ){
            sz = " ".repeat(5-sz.length())+sz;
        }

        return sz;
    }

    /**
     * Сортировка по умолчанию:
     * <ol>
     *     <li> Не null идут первыми
     *     <li> Каталоги идут первыми
     *     <li> Файлы/каталог сравниваются по именам
     * </ol>
     */
    public final Comparator<Path> defaultSort = (a,b) -> {
        if( a==null && b==null )return 0;
        if( a!=null && b==null )return -1;
        if( a==null )return 1;

        boolean a_dir = Files.isDirectory(a);
        boolean b_dir = Files.isDirectory(b);
        if( a_dir && !b_dir )return -1;
        if( !a_dir && b_dir )return 1;

        String a_name = getName(a);
        String b_name = getName(b);
        return a_name.compareTo(b_name);
    };
}

File: jtfm-scala/src/main/scala/xyz/cofe/jtfm/wid/cmpt/FilesTable.scala
package xyz.cofe.jtfm.wid.cmpt

import java.nio.file.Path
import java.nio.file.Files
import java.nio.file.attribute.FileTime
import java.time.ZoneId
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter
import java.time.format.DateTimeFormatterBuilder
import org.slf4j.LoggerFactory
import java.nio.file.attribute.PosixFileAttributes
import java.nio.file.LinkOption
import java.nio.file.attribute.PosixFilePermission

class FilesTable extends Table[Path] {
  import FilesTable.log
}

object FilesTable {  
  val log = LoggerFactory.getLogger(classOf[FilesTable])

  /** 
   * Текстовое представление размера файла
   */
  def humanSize(size:Long):String = {
    if size<0 then
      size.toString
    else
      if size < 1024 then
        size.toString
      else if size < 1024*1024 then
        // K
        (size / 1024)+"k"
      else if size < 1024*1024*1024 then
        // M
        (size / 1024 / 1024)+"m"
      else
        // G
        (size / 1024 / 1024 / 1024 )+"g"
  }

  val yyyyMMdd = DateTimeFormatter.ofPattern("yyyy-MM-dd")
  val HHmmss   = DateTimeFormatter.ofPattern("hh:mm:ss")

  def dateTimeShort(ft:FileTime):String = {
    val dt = ft.toInstant.atZone(ZoneId.systemDefault)
    val now = LocalDateTime.now
    if dt.getYear == now.getYear then
      HHmmss.format(dt)
    else
      yyyyMMdd.format(dt)
  }

  /**
   * Колонки файлов
   */
  object columns {
    /**
     * Имя файла
     */
    val fileName = Column.create[Path,String](
      "name", 
      path=>{ 
        val pref = 
          if Files.isSymbolicLink(path) then "~"
          else if Files.isDirectory(path) then "/"
          else " "
        pref + path.getFileName.toString
      },
      _.toString
    )    

    /**
     * Размер файла
     */
    val size = {
      val c = Column.create[Path,Long](
        "size", 
        path => {
          if( Files.exists(path) && Files.isRegularFile(path) ){
            try {
              Files.size(path)
            } catch {
              case err:Throwable => 
                log.warn(s"can't read size of $path",err)
                -2L
            }
          }else{
            log.debug(s"not file or exists $path")
            -1L
          }
        },
        size => humanSize(size)
      )
      c.width.prefect = Some(6)
      c
    }

    /**
     * Дата модификации
     */
    val latModifiedTime = {
      val c = Column.create[Path,FileTime](
        "modified",
        path => {
          try { Files.getLastModifiedTime(path) } catch {
            case err : Throwable => 
              log.warn(s"can't read getLastModifiedTime of $path",err)
              null
          }
        },
        nullableFileTime => if nullableFileTime==null then "?" else dateTimeShort(nullableFileTime)
      )
      c.width.prefect = Some(10)
      c
    }

    def posixPerm(p:PosixFileAttributes):String =
      (if p.permissions.contains(PosixFilePermission.GROUP_READ) then "r" else "-") +
      (if p.permissions.contains(PosixFilePermission.GROUP_WRITE) then "w" else "-") +
      (if p.permissions.contains(PosixFilePermission.GROUP_EXECUTE) then "x" else "-") +
      (if p.permissions.contains(PosixFilePermission.OWNER_READ) then "r" else "-") +
      (if p.permissions.contains(PosixFilePermission.OWNER_WRITE) then "w" else "-") +
      (if p.permissions.contains(PosixFilePermission.OWNER_EXECUTE) then "x" else "-") +
      (if p.permissions.contains(PosixFilePermission.OTHERS_READ) then "r" else "-") +
      (if p.permissions.contains(PosixFilePermission.OTHERS_WRITE) then "w" else "-") +
      (if p.permissions.contains(PosixFilePermission.OTHERS_EXECUTE) then "x" else "-")

    implicit val posixOptOrd:Ordering[Option[PosixFileAttributes]] = 
      (a,b) => (a,b) match {
        case (None,None) => 0
        case (Some(_),None) => 0
        case (None,Some(_)) => 0
        case (Some(x),Some(y)) =>
          if x.group.getName == y.group.getName then x.group.getName.compareTo(y.group.getName)
          else if x.owner.getName == y.owner.getName then x.owner.getName.compareTo(y.owner.getName)
          else posixPerm(x).compareTo( posixPerm(y) )
      }

    def posixColumn(posix2str:PosixFileAttributes=>String):Column[Path,Option[PosixFileAttributes]]=
      Column.create[Path,Option[PosixFileAttributes]](
        "perm",
        path => {
          try {
            Some(Files.readAttributes(path,classOf[PosixFileAttributes],LinkOption.NOFOLLOW_LINKS))
          } catch {
            case err: Throwable =>
              log.warn(s"can't readAttributes PosixFileAttributes of $path",err)
              None
          }
        },
        posixOpt => posixOpt match {
          case None => "?"
          case Some(posix) => posix2str(posix)
        }
      )

    val chmod = {
      val c = posixColumn(
        posix => posixPerm(posix)
      ).copy(name = "perm")
      c.width.prefect = Some(9)
      c
    }
  }

  object sort {
    lazy val defaultSort = combine(List(
      dirFirst, byName
    ))

    val dirFirst:(Path,Path)=>Int = (a,b) => {
      val aDir = try { Files.isDirectory(a) } catch { 
        case e:Throwable => log.warn(s"can't read attr(isDirectory) of $a",e); false
      }
      val bDir = try { Files.isDirectory(b) } catch { 
        case e:Throwable => log.warn(s"can't read attr(isDirectory) of $b",e); false
      }
      aDir == bDir match {
        case true => 0
        case _ => if aDir then -1 else 1
      }
    }

    val byName:(Path,Path)=>Int = (a,b) => {
      val aName = a.getFileName.toString
      val bName = b.getFileName.toString
      if aName==null && bName==null then
        0
      else if aName!=null && bName==null then
        -1
      else if aName==null && bName!=null then
        1
      else
        aName.compareTo(bName)
    }

    def combine( sorts:Seq[(Path,Path)=>Int] ):(Path,Path)=>Int = (fa,fb)=> {
      if sorts.isEmpty then
        0
      else
        var cmpRes = 0
        sorts.foldLeft( 0 )( (a,b) => {
          if a==0 then
            b(fa,fb)
          else 
            a
        })
    }

    def inverse( cmpFn:(Path,Path)=>Int ):(Path,Path)=>Int = (a,b) => cmpFn(a,b) * -1
  }
}
